matchmake-extension/database/get_official_communities.go

```diff
@@ -0,0 +1,137 @@
+	pqextended "github.com/PretendoNetwork/pq-extended"
+)
+
+// GetOfficialCommunities returns the official communities based on the given parameters
+func GetOfficialCommunities(manager *common_globals.MatchmakingManager, sourcePID types.PID, isAvailableOnly bool, resultRange types.ResultRange) ([]match_making_types.PersistentGathering, *nex.Error) {
+	persistentGatherings := make([]match_making_types.PersistentGathering, 0)
+	currentTime := time.Now().UTC()
+	rows, err := manager.Database.Query(`SELECT
+		g.id,
+		g.owner_pid,
+		g.host_pid,
+		g.min_participants,
+		g.max_participants,
+		g.participation_policy,
+		g.policy_argument,
+		g.flags,
+		g.state,
+		g.description,
+		pg.community_type,
+		pg.password,
+		pg.attribs,
+		pg.application_buffer,
+		pg.participation_start_date,
+		pg.participation_end_date
+		FROM matchmaking.gatherings AS g
+		INNER JOIN matchmaking.persistent_gatherings AS pg ON g.id = pg.id
+		WHERE
+		g.registered=true AND
+		g.type='PersistentGathering' AND
+		pg.community_type=2 AND
+		(CASE WHEN $1 THEN $2 BETWEEN pg.participation_start_date AND pg.participation_end_date ELSE true END)
+		LIMIT $3 OFFSET $4`,
+		isAvailableOnly,
+		currentTime,
+		uint32(resultRange.Length),
+		uint32(resultRange.Offset),
+	)
+	if err != nil {
+		return nil, nex.NewError(nex.ResultCodes.Core.Unknown, err.Error())
+	}
+
+	for rows.Next() {
+		var gatheringID uint32
+		var ownerPID uint64
+		var hostPID uint64
+		var minimumParticipants uint16
+		var maximumParticipants uint16
+		var participationPolicy uint32
+		var policyArgument uint32
+		var flags uint32
+		var state uint32
+		var description string
+		var communityType uint32
+		var password string
+		var resultAttribs []uint32
+		var applicationBuffer []byte
+		var resultParticipationStartDate time.Time
+		var resultParticipationEndDate time.Time
+
+		err = rows.Scan(
+			&gatheringID,
+			&ownerPID,
+			&hostPID,
+			&minimumParticipants,
+			&maximumParticipants,
+			&participationPolicy,
+			&policyArgument,
+			&flags,
+			&state,
+			&description,
+			&communityType,
+			&password,
+			pqextended.Array(&resultAttribs),
+			&applicationBuffer,
+			&resultParticipationStartDate,
+			&resultParticipationEndDate,
+		)
+		if err != nil {
+			common_globals.Logger.Critical(err.Error())
+			continue
+		}
+
+		resultPersistentGathering := match_making_types.NewPersistentGathering()
+		resultPersistentGathering.ID = types.NewUInt32(gatheringID)
+
+		resultMatchmakeSessionCount, nexError := GetPersistentGatheringSessionCount(manager, uint32(resultPersistentGathering.ID))
+		if nexError != nil {
+			common_globals.Logger.Critical(nexError.Error())
+			continue
+		}
+
+		resultParticipationCount, nexError := GetPersistentGatheringParticipationCount(manager, uint32(resultPersistentGathering.ID), uint64(sourcePID))
+		if nexError != nil {
+			common_globals.Logger.Critical(nexError.Error())
+			continue
+		}
```

Comment:
Same comment as before about reducing the DB calls (our DB gets pretty hammered as is). I won't leave a new comment for each instance, only doing the 2